sklearn/datasets/openml.py

```diff
@@ -263,6 +265,58 @@ def _convert_arff_data(arff_data, col_slice_x, col_slice_y, shape=None):
         raise ValueError('Unexpected Data Type obtained from arff.')
 
 
+def _feature_to_dtype(feature):
+    """Map feature to dtype for pandas DataFrame
+    """
+    if feature["data_type"] == "string":
+        return object
+    elif feature["data_type"] == "nominal":
+        return 'category'
+    # only numeric, integer, real are left
+    elif (feature["number_of_missing_values"] != "0" or
+          feature["data_type"] in ["numeric", "real"]):
+        return np.float64
+    elif feature["data_type"] == "integer":
+        return np.int64
+    raise ValueError("Unsupported feature: {}".format(feature))
+
+
+def _convert_arff_data_dataframe(arrf_data, all_columns, features_dict):
+    """Convert the ARFF object into a pandas DataFrame.
+
+    Parameters
+    ----------
+    arff_data : list or dict
+        as obtained from liac-arff object
+
+    all_columns : list
+        columns to return
+
+    features_dict : OrderedDict
+        map from feature to feature info from openml. This includes
+        columns that are not ignored.
+
+    Returns
+    -------
+    df : pd.DataFrame
+    """
+    check_pandas_support('fetch_openml with return_frame=True')
+    import pandas as pd
+
+    df = pd.DataFrame(arrf_data['data'], columns=list(features_dict.keys()),
```

Comment:
`arff_data['data']` is a generator, right? Does that work with all supported versions of Pandas? Can you make sure that it does not just do `list(data)` and explode the memory usage?

Reply:
`pd.Dataframe` does call `list(data)`, which we need to avoid.

> Otherwise we can pass it to pd.DataFrame.from_records()

Just tested `from_records` out and it seems to work. :)